src/compile/model.ts

```diff
@@ -41,24 +41,39 @@ export interface Component {
   mark: VgMarkGroup[];
 }
 
-class NameMap {
+export class NameMap {
   private _nameMap: Dict<string>;
 
   constructor() {
     this._nameMap = {} as Dict<string>;
   }
 
   public rename(oldName: string, newName: string) {
+    if (this._nameMap[newName] === oldName) {
+      console.error('Cannot rename ' + oldName + ' to ' + newName);
+      return;
+    }
+
+    if (newName in this._nameMap) {
+      // since we already have a new name for the target, we need to change it
+      newName = this._nameMap[newName];
+    } else {
+      // update existing renames
```

Comment:
The most optimized way to implement this is to update the map inside get before returning value. 

This requires searching over the map every time.